src/lib/foreign-field.ts

```diff
@@ -0,0 +1,496 @@
+      if (x instanceof ForeignField) {
+        this.value = x.value;
+        return;
+      }
+      // ForeignFieldVar
+      if (Array.isArray(x)) {
+        this.value = x;
+        return;
+      }
+      // constant
+      this.value = ForeignFieldVar.fromBigint(mod(BigInt(x), p));
+    }
+
+    /**
+     * Coerce the input to a {@link ForeignField}.
+     */
+    static from(x: ForeignField | ForeignFieldVar | bigint | number | string) {
+      if (x instanceof ForeignField) return x;
+      return new ForeignField(x);
+    }
+
+    /**
+     * Checks whether this field element is a constant.
+     *
+     * See {@link FieldVar} tp understand constants vs variables.
+     */
+    isConstant() {
+      let [, ...limbs] = this.value;
+      return limbs.every(FieldVar.isConstant);
+    }
+
+    /**
+     * Convert this field element to a constant.
+     *
+     * See {@link FieldVar} to understand constants vs variables.
+     *
+     * **Warning**: This function is only useful in `Provable.witness()` or `Provable.asProver()` blocks,
+     * that is, in situations where the prover computes a value outside provable code.
+     */
+    toConstant(): ForeignField {
+      let [, ...limbs] = this.value;
+      let constantLimbs = mapTuple(limbs, (l) =>
+        FieldVar.constant(FieldVar.toConstant(l))
+      );
+      return new ForeignField([0, ...constantLimbs]);
+    }
+
+    /**
+     * Convert this field element to a bigint.
+     */
+    toBigInt() {
+      return ForeignFieldVar.toBigint(this.value);
+    }
+
+    /**
+     * Assert that this field element lies in the range [0, p),
+     * where p is the foreign field modulus.
+     */
+    assertValidElement() {
+      if (this.isConstant()) return;
+      Snarky.foreignField.assertValidElement(this.value, pMl);
+    }
+
+    // arithmetic with full constraints, for safe use
+
+    /**
+     * Finite field addition
+     *
+     * ```ts
+     * x.add(2); // x + 2 mod p
+     * ```
+     */
+    add(y: ForeignField | bigint | number) {
+      if (this.isConstant() && isConstant(y)) {
+        let z = mod(this.toBigInt() + toFp(y), p);
+        return new ForeignField(z);
+      }
+      let z = Snarky.foreignField.add(this.value, toVar(y), pMl);
+      return new ForeignField(z);
+    }
+
+    /**
+     * Finite field negation
+     *
+     * ```ts
+     * x.neg(); // -x mod p = p - x
+     * ```
+     */
+    neg() {
+      if (this.isConstant()) {
+        let x = this.toBigInt();
+        let z = x === 0n ? 0n : p - x;
+        return new ForeignField(z);
+      }
+      let z = Snarky.foreignField.sub(ForeignFieldVar[0], this.value, pMl);
+      return new ForeignField(z);
+    }
+
+    /**
+     * Finite field subtraction
+     *
+     * ```ts
+     * x.sub(1); // x - 1 mod p
+     * ```
+     */
+    sub(y: ForeignField | bigint | number) {
+      if (this.isConstant() && isConstant(y)) {
+        let z = mod(this.toBigInt() - toFp(y), p);
```

Comment:
yeah, bigint subtraction works and can be negative, and z will always be positive. the modulo operation `%` in JS can return a negative number, that's why we use our own `mod()` function which is written so that the result is always in [0, p): https://github.com/o1-labs/snarkyjs-bindings/blob/6b069e49b9c80b6004262333e83c33047a8293eb/crypto/finite_field.ts#L26-L30